### ISSUE TITLE
"There is no validation function for format 'int16'"

### ISSUE THREAD

**Issue by glen-84** [swagger-api/swagger-editor#612]
According to the docs:

> However, the format property is an open string-valued property, and can have any value to support documentation needs.

... therefore it should not be validated.

**Comment by mohsen1** [comment]
That's an schema bug. @webron can you confirm?

**Comment by webron** [comment]
@mohsen1 - as far as I can tell, `format` is defined as `type: string` throughout the schema with no additional restrictions, so it doesn't seem to be a schema bug in this case.

**Comment by mohsen1** [comment]
Right. Maybe @glen-84 put `int16` in `type` instead of `format` by mistake. I can't reproduce it:

```yaml
---
swagger: '2.0'
info:
  version: 0.0.0
  title: Simple API
paths:
  /:
    get:
      parameters:
        - name: parameters
          in: query
          description: description
          type: string
          format: int16
      responses:
        200:
          description: OK
          schema:
            $ref: '#/definitions/Name'
definitions:
  Name:
    type: string
    format: int16
```

**Comment by glen-84** [comment]
@mohsen1 Add `default: 123`, and you will see the error.

**Comment by mohsen1** [comment]
I'm assuming you have changed the type to `number`. 

With that, `int16` is throwing an error

```
      parameters:
        - name: parameters
          in: query
          description: description
          type: number
          format: int16
          default: 123
```

but `int32` is fine. @webron is this a bug?

**Comment by whitlockjc** [comment]
`int16` is not a valid format.

**Comment by webron** [comment]
@whitlockjc -

> However, the format property is an open string-valued property, and can have any value to support documentation needs. Formats such as "email", "uuid", etc., can be used even though they are not defined by this specification. Types that are not accompanied by a format property follow their definition from the JSON Schema (except for file type which is defined above).

So, for example, whether true or not, `JeremyIsAwesome` is a valid format.

@mohsen1 - the type has no affect on the validity of a `format`. How or whether the value is validated though is a different matter.

**Comment by whitlockjc** [comment]
But `int16` is not a valid JSON Schema format either.  So any JSON Schema validator will bomb on this without registering a custom validator.

**Comment by whitlockjc** [comment]
In fact, I just tested this here https://json-schema-validator.herokuapp.com/ with the following: 

```js
{
  "type": "number",
  "format": "int16"
}
```

And got when validating the schema:

```
[ {
  "level" : "warning",
  "schema" : {
    "loadingURI" : "#",
    "pointer" : ""
  },
  "domain" : "validation",
  "keyword" : "format",
  "message" : "format attribute \"int16\" not supported",
  "attribute" : "int16"
} ]
```

However, using a valid *(JSON Schema format or Swagger format)* format works just fine.  Here is an example:

```js
{
  "type": "string",
  "format": "email"
}
```

**Comment by whitlockjc** [comment]
Of course, most JSON Schema validators have ways to register custom validators...which is how the sway validator used by swagger-editor handle's Swagger's formats.  But registering custom, non-Swagger and non-JSON Schema validators, seemed to be outside the scope of things so nothing was ever supported in swagger-tools or sway.

**Comment by webron** [comment]
I'm afraid that's inaccurate as well. From http://json-schema.org/latest/json-schema-validation.html#anchor106 - 

> Implementations MAY support the "format" keyword. Should they choose to do so:

>    they SHOULD implement validation for attributes defined below;

>    they SHOULD offer an option to disable validation for this keyword.

> **Implementations MAY add custom format attributes. Save for agreement between parties, schema authors SHALL NOT expect a peer implementation to support this keyword and/or custom format attributes.**

That suggests that if the formats are unknown, they can (and should) be ignored.

**Comment by whitlockjc** [comment]
I'm telling you what the "official" validator does and what the validator libraries I've used do.  I'm not making it up, try the website I linked to and/or the libraries.

**Comment by whitlockjc** [comment]
jjv, z-schema, the Java one powering https://json-schema-validator.herokuapp.com/ all work the same and fail when you use an unknown format.  All of them also have APIs to register custom formatters to work around this.  Here is an example for z-schema: https://github.com/zaggino/z-schema#register-a-custom-format

**Comment by webron** [comment]
It's an implementation and they may have chosen to not turn off validation of unknown formats. The JSON Schema specification explicitly says they are allowed, and in fact, JSON Schema's JSON Schema is non-restrictive - https://github.com/json-schema/json-schema/blob/master/draft-04/schema#L140.

And just because they all fail, doesn't mean they don't all have a bug in them. I didn't write that specification :)

**Comment by whitlockjc** [comment]
This is a z-schema bug.  We should report it there.  I guess we could report it to sway as well.

**Comment by whitlockjc** [comment]
I've deleted my comments.  I made a few of them not realizing the implementation linked to was giving a warning and not an error.  Anyway, the JSON Schema validator swagger-editor uses is z-schema and I'm filing a bug there to allow us to not fail on unknown format.

**Comment by whitlockjc** [comment]
Done...I'll submit a PR there and when that is released, I'll update sway so that swagger-editor can be itself updated.

**Comment by webron** [comment]
Thanks, @whitlockjc, much appreciated.

**Comment by whitlockjc** [comment]
https://github.com/zaggino/z-schema/pull/133  As soon as it's pulled an z-schema releases, we'll update sway and swagger-editor.

**Comment by mohsen1** [comment]
That's a long dependency chain! 
`z-schema->sway->sway-worker->swagger-editor`

So basically we should not throw an error for this format value.

**Comment by whitlockjc** [comment]
That's what the PR above does.  Once it's supported, I will update the z-schema version in sway and configure sway to use this new feature.

**Comment by mohsen1** [comment]
This was fixed by b6714d1

**Comment by glen-84** [comment]
Thanks. :+1:
